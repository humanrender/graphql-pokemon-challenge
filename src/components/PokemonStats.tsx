import React, { useEffect, useState } from 'react';
import styled from 'styled-components';

import { Button, Title } from '../styled/layout';
import { COLORS, SPACING } from '../styled/shared';
import { Pokemon, PokemonMove } from '../types';

export interface PokemonStatsInterface {
  pokemon?: Pokemon | null;
  loading?: boolean;
  onPokemonSelected: (pokemon: Pokemon, selectedMoves: PokemonMove[]) => void;
}

const Wrapper = styled.div`
  display: flex;
  height: 100%;
  gap: ${SPACING * 2}px;
`;

const PokemonDescription = styled.div`
  max-width: ${SPACING * 18}px;
  flex: 1 0 auto;
`;

const PokemonImage = styled.img`
  height: ${SPACING * 12}px;
`;

const PokemonName = styled.p`
  color: ${COLORS.blue};
  text-transform: uppercase;
  margin-top: ${SPACING}px;
  margin-bottom: ${SPACING}px;
`;

const Subtitle = styled.h3`
  color: ${COLORS.blue};
  text-transform: uppercase;
  display: block;
  margin-top: 0;
  margin-bottom: ${SPACING}px;
  text-align: center;
  font-size: 14px;
`;

const Stat = styled.p`
  color: ${COLORS.yellow};
  margin-bottom: ${SPACING}px;
  width: 50%;
  text-transform: uppercase;
  font-size: 14px;
  text-align: right;
`;

const StatValue = styled.span`
  color: ${COLORS.blue};
  font-size: 20px;
`;

const StatsWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

const MoveListGroups = styled.ul`
  list-style: none;
  padding: 0;
  margin: 0;
  overflow: auto;
  min-width: ${SPACING * 23}px;
  max-height: ${SPACING * 32}px;
`;

const MoveList = styled.ul`
  list-style: none;
  padding: 0;
  margin: 0;
  color: blue;
`;

const SelectedMoves = styled.ol`
  display: flex;
  gap: ${SPACING}px;
  flex-wrap: wrap;
  list-style: none;
`;

const SelectedMove = styled.li`
  border: 1px solid ${COLORS.blue};
  box-sizing: border-box;
  width: calc(50% - ${SPACING / 2}px);
`;

const MAX_MOVES = 4;

export const PokemonStats = ({
  loading,
  pokemon,
  onPokemonSelected,
}: PokemonStatsInterface) => {
  const stats: Pokemon['stats'] = pokemon?.stats || ({} as Pokemon['stats']);
  const [selectedMoves, setSelectedMoves] = useState<PokemonMove[]>([]);
  useEffect(() => {
    setSelectedMoves([]);
  }, [pokemon]);
  return (
    <Wrapper>
      <PokemonDescription>
        {loading && <p>Loading</p>}
        {pokemon && (
          <>
            {!loading && pokemon?.name && (
              <>
                <PokemonImage src={pokemon.image} />
                <PokemonName>{pokemon.name}</PokemonName>
                <Button
                  secondary
                  onClick={() => onPokemonSelected(pokemon, selectedMoves)}
                >
                  Save Pokemon
                </Button>
              </>
            )}
          </>
        )}
      </PokemonDescription>
      <div>
        <Subtitle>Stats</Subtitle>
        <StatsWrapper>
          <Stat>
            Speed <StatValue>{stats.speed || '-'}</StatValue>
          </Stat>
          <Stat>
            Special-Defense <StatValue>{stats.specialDefense || '-'}</StatValue>
          </Stat>
          <Stat>
            Special-Attack <StatValue>{stats.specialAttack || '-'}</StatValue>
          </Stat>
          <Stat>
            Defense <StatValue>{stats.defense || '-'}</StatValue>
          </Stat>
          <Stat>
            Attack <StatValue>{stats.attack || '-'}</StatValue>
          </Stat>
          <Stat>
            HP <StatValue>{stats.hp || '-'}</StatValue>
          </Stat>
        </StatsWrapper>
        <Subtitle>Selected moves</Subtitle>
        <SelectedMoves>
          {selectedMoves.map((move) => (
            <SelectedMove
              key={move.name}
              onClick={() => {
                const newMoves = [...selectedMoves].filter(
                  (newMove) => newMove.name !== move.name
                );
                setSelectedMoves(newMoves);
              }}
            >
              <small>{move.learnMethod}</small> <p>{move.name}</p>
            </SelectedMove>
          ))}
        </SelectedMoves>
      </div>
      <MoveListGroups>
        {Object.entries(pokemon?.availableMoves || {}).map(
          ([learnMethod, moves]) => (
            <li key={learnMethod}>
              <Title>{learnMethod}</Title>
              <MoveList>
                {moves.map((move) => (
                  <li
                    key={move.name}
                    onClick={() => {
                      let newMoves = [...selectedMoves];
                      if (
                        newMoves.find((newMove) => newMove.name === move.name)
                      ) {
                        newMoves = newMoves.filter(
                          (newMove) => newMove.name !== move.name
                        );
                      } else if (Object.keys(newMoves).length < MAX_MOVES) {
                        newMoves.push(move);
                      }
                      setSelectedMoves(newMoves);
                    }}
                  >
                    {move.name}
                  </li>
                ))}
              </MoveList>
            </li>
          )
        )}
      </MoveListGroups>
    </Wrapper>
  );
};
