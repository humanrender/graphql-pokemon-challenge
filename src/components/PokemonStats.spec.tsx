import * as React from 'react';
import renderer from 'react-test-renderer';

import { PokemonStats } from './PokemonStats';

describe('PokemonStats', () => {
  describe('without pokemon', () => {
    it('should render the default view', () => {
      const onPokemonSelected = jest.fn();
      const tree = renderer
        .create(
          <PokemonStats pokemon={null} onPokemonSelected={onPokemonSelected} />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('with a pokemon', () => {
    it('should render the default view', () => {
      const onPokemonSelected = jest.fn();
      const tree = renderer
        .create(
          <PokemonStats
            pokemon={{
              id: '143',
              name: 'snorlax',
              image:
                'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/143.png',
              type: 'normal',
              abilities: ['immunity', 'thick-fat', 'gluttony'],
              availableMoves: {
                machine: [{ name: 'defense-curl', learnMethod: 'machine' }],
              },
              stats: {
                hp: 123,
                defense: 234,
                attack: 345,
                speed: 456,
                specialAttack: 567,
                specialDefense: 678,
              },
            }}
            onPokemonSelected={onPokemonSelected}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('when loading', () => {
    it('should render the loading view', () => {
      const onPokemonSelected = jest.fn();
      const tree = renderer
        .create(
          <PokemonStats
            pokemon={null}
            onPokemonSelected={onPokemonSelected}
            loading
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
