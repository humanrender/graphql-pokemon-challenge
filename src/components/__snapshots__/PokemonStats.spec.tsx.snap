// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonStats when loading should render the loading view 1`] = `
<div
  className="sc-htpNat iTxZMQ"
>
  <div
    className="sc-bxivhb fjbmad"
  >
    <p>
      Loading
    </p>
  </div>
  <div>
    <h3
      className="sc-bZQynM hXicKp"
    >
      Stats
    </h3>
    <div
      className="sc-dnqmqq eWlhsD"
    >
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Speed 
        <span
          className="sc-htoDjs cOUezd"
        >
          -
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Special-Defense 
        <span
          className="sc-htoDjs cOUezd"
        >
          -
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Special-Attack 
        <span
          className="sc-htoDjs cOUezd"
        >
          -
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Defense 
        <span
          className="sc-htoDjs cOUezd"
        >
          -
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Attack 
        <span
          className="sc-htoDjs cOUezd"
        >
          -
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        HP 
        <span
          className="sc-htoDjs cOUezd"
        >
          -
        </span>
      </p>
    </div>
    <h3
      className="sc-bZQynM hXicKp"
    >
      Selected moves
    </h3>
    <ol
      className="sc-gqjmRU hxwYYq"
    />
  </div>
  <ul
    className="sc-iwsKbI ftKUSZ"
  />
</div>
`;

exports[`PokemonStats with a pokemon should render the default view 1`] = `
<div
  className="sc-htpNat iTxZMQ"
>
  <div
    className="sc-bxivhb fjbmad"
  >
    <img
      className="sc-ifAKCX fhCdPi"
      src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/143.png"
    />
    <p
      className="sc-EHOje dWaoIP"
    >
      snorlax
    </p>
    <button
      className="sc-bwzfXH kfFXpC"
      onClick={[Function]}
    >
      Save Pokemon
    </button>
  </div>
  <div>
    <h3
      className="sc-bZQynM hXicKp"
    >
      Stats
    </h3>
    <div
      className="sc-dnqmqq eWlhsD"
    >
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Speed 
        <span
          className="sc-htoDjs cOUezd"
        >
          456
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Special-Defense 
        <span
          className="sc-htoDjs cOUezd"
        >
          678
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Special-Attack 
        <span
          className="sc-htoDjs cOUezd"
        >
          567
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Defense 
        <span
          className="sc-htoDjs cOUezd"
        >
          234
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Attack 
        <span
          className="sc-htoDjs cOUezd"
        >
          345
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        HP 
        <span
          className="sc-htoDjs cOUezd"
        >
          123
        </span>
      </p>
    </div>
    <h3
      className="sc-bZQynM hXicKp"
    >
      Selected moves
    </h3>
    <ol
      className="sc-gqjmRU hxwYYq"
    />
  </div>
  <ul
    className="sc-iwsKbI ftKUSZ"
  >
    <li>
      <h2
        className="sc-bdVaJa caOGxZ"
      >
        machine
      </h2>
      <ul
        className="sc-gZMcBi cOcRph"
      >
        <li
          onClick={[Function]}
        >
          defense-curl
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`PokemonStats without pokemon should render the default view 1`] = `
<div
  className="sc-htpNat iTxZMQ"
>
  <div
    className="sc-bxivhb fjbmad"
  />
  <div>
    <h3
      className="sc-bZQynM hXicKp"
    >
      Stats
    </h3>
    <div
      className="sc-dnqmqq eWlhsD"
    >
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Speed 
        <span
          className="sc-htoDjs cOUezd"
        >
          -
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Special-Defense 
        <span
          className="sc-htoDjs cOUezd"
        >
          -
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Special-Attack 
        <span
          className="sc-htoDjs cOUezd"
        >
          -
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Defense 
        <span
          className="sc-htoDjs cOUezd"
        >
          -
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        Attack 
        <span
          className="sc-htoDjs cOUezd"
        >
          -
        </span>
      </p>
      <p
        className="sc-gzVnrw bEUQaE"
      >
        HP 
        <span
          className="sc-htoDjs cOUezd"
        >
          -
        </span>
      </p>
    </div>
    <h3
      className="sc-bZQynM hXicKp"
    >
      Selected moves
    </h3>
    <ol
      className="sc-gqjmRU hxwYYq"
    />
  </div>
  <ul
    className="sc-iwsKbI ftKUSZ"
  />
</div>
`;
